Added assumption methods to Sum and Product

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #17424

#### Brief description of what is fixed or changed
Inspired by #17412 I added assumptions on Sum and Products. This solves some of the cases handled in #17412, but not all, so both PRs are worthwhile. This may have positive implications elsewhere as well, but not really clear where.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* concrete
   * Better assumptions on `Sum` and `Product`.
   * Empty summations leads to zero result.
<!-- END RELEASE NOTES -->

oscargus committed:
Added assumption methods to Sum and Product

oscarbenjamin commented on sympy/concrete/products.py:
```diff
@@ -204,6 +204,17 @@ def _eval_is_zero(self):
+
+    def _eval_is_positive(self):
+        if self.function.is_positive:
+            return True
```
If this is an infinite product then can it not be zero or infinite even if the function is positive e.g.:
```julia
In [2]: Product(S(1)/2, (x, 1, oo)).doit()                                                                                        
Out[2]: 0

In [3]: Product(2, (x, 1, oo)).doit()                                                                                             
Out[3]: ∞
```
I think that a positive function can imply that the product is extended_nonnegative in the infinite case though.

oscargus replied:
Makes sense. However if the limits are finite it shouldn't happen, right?

oscarbenjamin replied:
It won't happen if the limits are finite. I think it would be good to add an `is_finite_product` property that can be used conveniently in places like this.

oscargus replied:
I think `has_finite_limits` is my preferred name. Now, if there are no limits (as for Integrals, it makes sense to add this to `ExprWithLimits`), should one go for the assumptions on the variable or just return `None` if there are no explicit limits?

Btw, something similar happens for sums: `Sum(1/2, (x, 1, oo)).doit()` is `oo`.

oscarbenjamin replied:
I think that `has_finite_limits` makes sense. In the context of an integral without limits it should be None I think. That probably makes the most sense since in most contexts you would use the property the None case would lead to an inconclusive result. (It probably also makes sense for Integrals to have something like an is_definite property if they don't)

oscargus replied:
However, if you have a integration variable with is real, it should differ from extended_real. So right now, real leads to finite limits.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -174,6 +174,21 @@ def _eval_is_zero(self):
+        return self.function.is_extended_real
+
+    def _eval_is_positive(self):
+        return self.function.is_positive
```
Similar comments about infinite sums. Also we have to consider the empty sum here. I just went to show an example and got these:
```julia
In [8]: Sum(1, (x, 0, 0)).doit()                                                                                                  
Out[8]: 1

In [9]: Sum(1, (x, 0, -1)).doit()                                                                                                 
Out[9]: 0

In [10]: Sum(1, (x, 0, -5)).doit()                                                                                                
Out[10]: -4
```
I don't know if I think those are correct though...

oscargus replied:
Inverted limits are clearly worse... I was under the impression that these were inverted, but I see it is in doit, not on initialization.

oscargus replied:
Should still be some cases where it is possible to determine that it works.

oscarbenjamin replied:
I still understand the convention for Sum limits well enough to answer this: how do I define an empty Sum?

oscargus replied:
I do not think that it is feasible, unless it is the same as having a negative difference of one.

oscargus committed:
Improve assumptions

oscargus commented:
I've updated, but it is not finished. Added a `has_reversed_limits` as well. (As it was shorter than `has_normal_ordered_limits`...)

The problem I've realized is that e.g. `Sum(1, (n, 2, M)).is_positive` will return None, even if `M` is positive integer, as it may be 1. Although in 99% of the cases, people would mean that it is >= 2. I'm wondering if one should give the benefit of doubt here and assume that people know what they are doing in these situations? So assume that symbolic bounds are always correctly ordered (unless something else says otherwise, like M being negative).

oscargus commented on sympy/concrete/products.py:
```diff
@@ -201,8 +201,24 @@ def term(self):
     def _eval_is_zero(self):
-        # a Product is zero only if its term is zero.
-        return self.term.is_zero
+        if self.has_finite_limits:
```
This should be corrected, but it was indeed a bug here earlier.

oscarbenjamin commented on sympy/concrete/products.py:
```diff
@@ -201,18 +201,23 @@ def term(self):
     def _eval_is_nonnegative(self):
-        if self.function.is_nonnegative:
+        if self.function.is_nonnegative and self.has_finite_limits:
+            return True
```
An `_eval_is_finite` method might be useful. The `_eval_is_nonnegative` method above for example could be automatically inferred from the extended version below if you had that. Similarly if you had `_eval_is_zero` then `_eval_is_positive` could be inferred.

oscargus replied:
Isn't that basically the same as `is_convergent`? I was considering it, but since it is quite computationally intensive, I didn't really want to involve it into the random evaluation of assumptions...

`is_zero` could just check for the argument being zero. Everything else require some sort of evaluation. So not sure how much it would help.

oscarbenjamin replied:
I would expect an `_eval_is_finite` method to be a quick check like: finite limits and finite summand -> return True, otherwise return None. This is essentially the same check that you have used to determine finiteness here.

I think there should be something like is_convergent but I don't think it should be a property. It should be a method with a docstring explaining that it is slow and potentially with options to control which algorithm is used etc.

oscarbenjamin replied:
`is_zero` would just check whether the argument is zero and the finiteness of the limits. It helps because you don't need to define all the `_eval_is` methods. Positive can be inferred from nonnegative and zero. In combination if you define extended_nonnegative, finite and zero then they can be used to infer all of positive, nonnegative, extended_positive etc. Also finite and zero are the most directly useful in core operations like division.

oscargus replied:
I think that some are easier than others to define. For `is_finite` one will need to check that there are no singularities in the range and that the bounds are finite. First one is not so simple for the general case. Hence, providing those that are reasonably easy to evaluate may be a better/working (initial) strategy as opposed to putting much effort into solving a potentially better check, taking more time.

oscargus replied:
I would assume that the assumption system continues checking other properties when `None` is obtained?

oscarbenjamin replied:
If the dummy symbols have assumptions that are consistent with the limits then function.is_finite and has_finite_limits are enough to imply that a Sum/Product is finite I think.

oscargus replied:
I see your point. That would indeed be a way to check it.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -178,16 +178,20 @@ def _eval_is_extended_real(self):
     def _eval_is_positive(self):
-        return self.function.is_positive
+        if self.has_finite_limits and self.has_reversed_limits is False:
+            return self.function.is_positive
```
Is the empty sum considered to have reversed limits?

oscargus replied:
I would guess so, yes. That is more or less the same behavior as in `doit`. One could probably check for that as well (which if nothing else would give ´_eval_is_zero` a bit more to work with).

I see the assumptions that they should not provide incorrect results. Then, they can always provide more or less conclusive results (I mean, one can always use doit to get the actual result...).

oscarbenjamin replied:
Exactly. It's a quick check that will return None if necessary. The user can use doit if they want.

oscargus replied:
Hence my reluctance to put too advanced evaluation in.

oscargus committed:
Minor fixes

oscarbenjamin commented:
> give the benefit of doubt here and assume that people know what they are doing

Resist that temptation! It's shortcuts like that that combine and multiply into hard to track down bugs in distant places.

This is also why I think the assumptions need to be checked on the dummy symbol. All of the `_eval_is_` methods you've written here depend on the dummy symbol assumptions and will fail if I do:
```julia
In [49]: n = Symbol('n', negative=True)                                                                                           

In [50]: Sum(n, (n, 1, 5)).is_positive
```

oscargus committed:
Improved handling of empty sequences

oscargus commented:
The example you provide will give an error on evaluation (now that #17259 is merged), so what is the correct behaviour here?
The question is really when that error should be visible?

oscargus commented:
It seems like summations and products over empty sequences were not properly handled, so I added a property `is_empty_sequence` (not really happy with the name). 

So this should now give the correct result (False):
`Product(0, (x, y, y-1)).is_zero`

I think that probably we should add max and min assumptions on symbols as well. That will help quite a bit, both here and in other situations. Right now, the only way to get `Sum(1, (n, 2, M)).is_positive`to return True is to write `Sum(1, (n, 2, 2+M)).is_positive` while it would be more natural to assume that M is not only positive but >= 2.

oscargus committed:
Fixed merge conflict and added tests and documentation

oscargus committed:
Added failing test

oscarbenjamin commented:
> I think that probably we should add max and min assumptions on symbols as well.

The new assumptions are hoped to gain the ability to say e.g. `Q(x>2)`. I'm not sure how useful the max/min assumptions would be. In most practical cases where you have a summation say from 1 to N there will be the possibility that it is an empty sum.

What the new assumptions should eventually make possible is saying: `summand = refine(summand, Q.integer(n) & Q(x>=a) & Q(x<=b))` though.

oscarbenjamin commented on sympy/concrete/expr_with_intlimits.py:
```diff
@@ -283,3 +284,57 @@ def reorder_limit(expr, x, y):
+            dif = lim[1] - lim[2]
+            eq = Eq(dif, 1)
+            if eq == True:
+                return True
```
How is Sum defined for non-integer limits like `Sum(f(n), (n, 1, S(3)/2))`?

oscargus replied:
No idea... I still haven't understood if it is an actual summation using those properties or if it simply applies the closed form equations. 

Here is a good set of examples which I think shows that it is a combination of approaches:

```
from sympy import *
m, n, x = symbols('m n x')
summation(sin(x), (x, S.Half, S.Half+2))
summation(sin(x), (x, S.Half, 3))
summation(x, (x, S.Half, 3))
summation(x, (x, S.Half, n))
summation(sin(x), (x, m, n))
```

Earlier on today, I realized that 
```
summation(x, (x, S.Half, 3+I))
```
gives a numerical result...

(Will get back with more comments.)

oscargus replied:
I think that the procedure is as follows:

If the difference between the upper and lower bound is a real integer it will try just replace the expression if there are less than 100 terms. (For reversed limits, it will flip it back first.)

If there is a symbolic solution is will just use that, without considering if the bounds are real or not and integer or not.

Which corresponds to the examples and the my rough understanding of the code.

Not sure if this is a very good approach though. Especially when it comes to allowing complex limits. As a consequence, conjugates are (currently) not propagated to the function. So at least it is consistent with the behaviour. Ping @asmeurer @smichr @jksuom : Do you have any opinions on if summations (and products) should be able to accept complex and/or non-integer arguments? I can see the point of being able to sum over arbitrary sets, but not really in the Sum_from^to-format.

jksuom replied:
I would only use integer limits. If something else is needed, sum over `a, a + 1, a + 2, ..., a + n`, for example, then I would rewrite the index as `a + i`.

oscargus replied:
Thanks @jksoum ! It will be a breaking change though if we force all limits to be integers:

https://github.com/sympy/sympy/blob/bfc42a2fc79deeec769761a2af6cadb1e621300e/sympy/concrete/tests/test_sums_products.py#L322-L327

But in my opinion it is worth it.

oscargus replied:
After playing around, I think one reason to allow non-integer limits is that they can happen after change_index-rewrites.

For example, trying to solve:
`Sum(1/(x**2+2*sqrt(2)*x + 2), (x, K, M))`

One way here is to instead solve
`Sum(1/(x**2+2*sqrt(2)*x + 2), (x, K, M)).change_index(x, x+sqrt(2)).simplify(doit=False)`

this doesn't work due to an error in try_lerchphi, but 
`Sum(1/(x**2+2*sqrt(2)*x + 2), (x, K, M)).change_index(x, x+sqrt(2)).simplify(doit=False).subs({M + sqrt(2) : M, K + sqrt(2) : K}).doit().subs({M : M + sqrt(2), K : K + sqrt(2)})`
does (which numerically give the same for all M and K I've tested). Now, there is no actual summation with non-integer limits involved, but ideally, `doit()` should be able to do these transformations automatically (but then, I guess that it can also do the subs so maybe not a big issue anyway...).

oscarbenjamin replied:
I think that non-integer limits can be okay as in something like:
```julia
In [3]: Sum(f(n), (n, -S(1)/2, S(1)/2)).doit()                                                                                    
Out[3]: f(-1/2) + f(1/2)
```
We should however require that the difference between the upper and lower limits be an integer so this is not okay:
```julia
In [4]: Sum(1, (n, 0, S(3)/2)).doit()                                                                                             
Out[4]: 5/2
```
The Sum should be required to have a nonnegative integer number of terms so if the summand is 1 then the result must always be an integer.

I think that we should also disallow "reversed" limits as in
```julia
In [7]: Sum(1, (n, 0, -2)).doit()                                                                                                 
Out[7]: -1
```
I think that if the upper limit is lower than the lower limit then the result should just be zero.

I propose the following equivalence:
```
Sum(f(n), (n, a, b)) = Sum(f(a+m), (m, 0, b-a)) = sum(f(a+m) for m in range(b-a+1))
```
This gives:
```julia
In [8]: def sumrange(expr, sym, a, b): 
   ...:     return sum(expr.subs(sym, a+m) for m in range(b-a+1)) 
   ...: 

In [13]: sumrange(f(n), n, 0, 3)                                                                                                  
Out[13]: f(0) + f(1) + f(2) + f(3)

In [14]: sumrange(f(n), n, S(1)/2, S(3)/2)                                                                                        
Out[14]: f(1/2) + f(3/2)

In [15]: sumrange(f(n), n, 0, -3)                                                                                                 
Out[15]: 0

In [16]: sumrange(f(n), n, 0, S(1)/2)                                                                                             
---------------------------------------------------------------------------
TypeError
```
This would mean that some general results should come back as Piecewise e.g.:
```julia
In [18]: Sum(1, (n, x, y)).doit()                                                                                                 
Out[18]: -x + y + 1

In [19]: Piecewise((1+y-x, x<=1+y))   # Correct answer
Out[19]: {-x + y + 1  for x ≤ y + 1
```
I think that it is okay to assume with symbolic limits that b-a is an integer if the assumptions don't justify it but if b-a is definitely not an integer then an error should be raised.

The docstring for Sum describes something different from what I suggest above but also doesn't seem to be consistent with current behaviour of actual Sum.

oscarbenjamin commented on sympy/concrete/expr_with_limits.py:
```diff
@@ -370,6 +370,111 @@ def _eval_subs(self, old, new):
+                var, a, b = lim
+                dif = b - a
+                if (dif >= 0) == False:
+                    return True
```
For `Sum(f(n), (n, oo, oo))` this will give invalid nan comparison

oscargus replied:
This type of comparison exists in quite a bit of code. (Integration for sure.)

oscargus replied:
It used to be like that but apparently changed now. I change this as well.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -171,9 +171,35 @@ def _eval_is_zero(self):
+
+    def _eval_is_nonnegative(self):
+        if self.is_zero:
+            return True
```
In some situations checking one assumption from another like this can lead to strange bugs to do with assumptions caching. What can happen is that the user asks `if e.is_zero` and then `_ask` ends up here because nonnegative=False implies zero=False. This asks for `is_zero` again but receives the `None` that was temporarily stored there and then you get different results from `_eval_is_nonnegative` than you would have had if the user had asked directly for `e.is_nonnegative`.

Because of the above I would say that a good rule of thumb is that `x._eval_is_Y` should avoid checking `x.is_Z`. Checking properties on `x.args` is fine though. Since the two lines above are not needed (`_ask` will check `_eval_is_zero` later if this returns inconclusive) I would just remove them.

oscargus replied:
I changed it so call _eval_is_zero, to avoid duplicating code. Not sure if it is always the case that all information in there can be derived from the other assumptions.

oscarbenjamin commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1164,6 +1167,132 @@ def test_exceptions():
+    i, j = symbols('i,j', positive=True, integer=True)
+    x, y = symbols('x,y', positive=True)
+    # Here, it must be rewritten as the assumption system cannot figure out
+    # that x**i*y**j is positive.
```
I don't understand what the above means but:
```julia
In [32]:     i, j = symbols('i,j', positive=True, integer=True) 
    ...:     x, y = symbols('x,y', positive=True)                                                                                 

In [33]: (x**i*y**j).is_positive                                                                                                  
Out[33]: True
```

oscargus replied:
OK! Then there is some other reason that it doesn't evaluate without the explicit rewrite.

oscargus commented:
If I understand correctly, the new assumptions only does that during refine?

I see huge benefits from it when it comes to relations.

Typically, things like `Integral(Piecewise((x, (x <= 1)), (2, True)), (x, 0, 1))` should be much easier to handle if one could add that x is between 0 and 1. (These show up in e.g. expressions that comes from Uniform variables, those actually have bounds when they are stochastic variables, but when turned into normal symbols they are removed).

Related to this PR: `Sum(1/(x-2), (x, 3, M)).is_finite` should give the correct answer more easily.

And this could happen while evaluating the expressions, not by getting a huge expression that can be refined afterwards.

I think this may be a quite likely thing for me to start working on. The question is really how complicated one should make it. If one should just support a single upper and lower bound or if one should have Unions of Intervals. The first seems simpler (but one will also need to consider how to deal with complex variables etc...).

oscargus committed:
Fixed merge failures

oscargus committed:
Fixed review comments

oscargus commented:
I've updated based on the review comments (and added some tests for is_finite and the problematic cases you pointed out).

A small drawback currently is that `doit` updates the assumptions on the summation/product symbols, but the assumption requests on the complete Sum/Product are determined based on the provided ones. Maybe one should have a function that updates the assumptions before performing the query? Drawback is that it will be quite time consuming. Or should one have a "shadowed" version that is created on instantiation which has the derived assumptions on the symbols and use that for assumptions? (Same "problem" with integrals, so maybe it is a later issue to deal with.)

oscarbenjamin committed:
Merge branch 'master' into sumproductassumptions

oscarbenjamin committed:
Remove trailing whitespace

oscarbenjamin commented:
Looks good. Merging
